tty/pc-tty/ttypc_virt.c

```diff
@@ -128,508 +152,490 @@ static void _ttypc_virt_scroll(ttypc_virt_t *virt)
+	if ((c >= 0x20) && (c <= 0x7f)) {               /* use GL if ch >= 0x20 */
 		if (!virt->ss)                              /* single shift G2/G3 -> GL ? */
-			*video = attrib | (*virt->GL)[ch - 0x20];
+			*video = attrib | (*virt->GL)[c - 0x20];
 		else {
-			*video = attrib | (*virt->GL)[ch - 0x20];
+			*video = attrib | (*virt->GL)[c - 0x20];
 			virt->ss = 0;
 		}
 	}
 	else {
 		virt->ss = 0;
 
-		if (ch >= 0xa0)                             /* display controls C1 */
-			*video = attrib | (*virt->GR)[ch - 0xa0];
-		else				                                /* display controls C0 */
-			*video = attrib | ch;
+		if (c >= 0xa0)                             /* display controls C1 */
+			*video = attrib | (*virt->GR)[c - 0xa0];
+		else                                       /* display controls C0 */
+			*video = attrib | c;
 	}
 }
 
 
-/* Emulator main entry */
-int ttypc_virt_sput(ttypc_virt_t *virt, uint8_t *s, int len)
+int ttypc_virt_sput(ttypc_virt_t *virt, char c)
 {
-	uint16_t ch;
-	int ret;
+	int ret = 1;
 
 	mutexLock(virt->mutex);
 
-	if (virt->rb == virt->rp)
-		virt->ready = 0;
+	/* always process control-chars in the range 0x00..0x1f, 0x7f !!! */
+	if (CTL_VALID(c)) {
+
+		switch (c) {
+		case 0x00:  /* NUL */
+		case 0x01:  /* SOH */
+		case 0x02:  /* STX */
+		case 0x03:  /* ETX */
+		case 0x04:  /* EOT */
+		case 0x05:  /* ENQ */
+		case 0x06:  /* ACK */
+		case 0x07:  /* BEL */
+		case 0x08:  /* BS  */
+			if (virt->col) {
+				virt->cur_offset--;
+				virt->col--;
+			}
+			ret = 0;
+			break;
 
-	ret = len;
-	while (len-- > 0) {
-		if ((ch = *(s++)) == 0)
+		case 0x09:  /* TAB */
+			while (virt->col < virt->maxcol - 1) {
+				virt->cur_offset++;
+				if (virt->tab_stops[++virt->col])
+					break;
+			}
 			break;
 
-		/* always process control-chars in the range 0x00..0x1f, 0x7f !!! */
-		if ((ch <= 0x1f) || (ch == 0x7f)) {
+		case 0x0a:  /* LF */
+		case 0x0b:  /* VT */
+		case 0x0c:  /* FF */
+#if 0
+			if (ttypc_virt_scrollback(virt)) {
+				extra = (svsp->cur_offset % svsp->maxcol) ? svsp->col : 0;
+				hal_memcpy(svsp->scrollback + svsp->scr_offset * svsp->maxcol,
+				           svsp->crtat + svsp->cur_offset - extra,
+				           svsp->maxcol * CHR);
+			}
+#endif
+			virt->cur_offset += virt->maxcol;
 
-			switch (ch) {
-			case 0x00:  /* NUL */
-			case 0x01:  /* SOH */
-			case 0x02:  /* STX */
-			case 0x03:  /* ETX */
-			case 0x04:  /* EOT */
-			case 0x05:  /* ENQ */
-			case 0x06:  /* ACK */
-				break;
-			case 0x07:  /* BEL */
-				break;
+			_ttypc_virt_scroll(virt);
+			break;
 
-			case 0x08:  /* BS */
+		case 0x0d:  /* CR */
+			virt->cur_offset -= virt->col;
+			virt->col = 0;
+			break;
 
-				if (virt->rp == virt->rb && ret) {
-					mutexUnlock(virt->mutex);
-					return 0;
-				}
+		case 0x0e:  /* SO */
+			virt->GL = &virt->G1;
+			break;
+		case 0x0f:  /* SI */
+			virt->GL = &virt->G0;
+			break;
 
-				if (virt->col) {
-					virt->cur_offset--;
-					virt->col--;
-				}
+		case 0x10:  /* DLE */
```

Comment:
You can compact all cases with _break;_ ending to the one block

Reply:
I'd like to leave it 'as is' and maintain case enumeration order.